docs/source/design/input_output_specifications.md

```diff
@@ -0,0 +1,692 @@
+    modality    : str                                 # e.g. "rna", "dna", "image"
+    semantic    : str                                 # Ontology reference (EDAM, OBI, GO, etc.)
+    organism    : str
+    version     : str = "1.0.0"                       # Increment when axes or encoding expectations change
+    axes        : Optional[List[Axis]] = None         # List of axes (e.g., cells, genes)
+    formats     : List[str] = ["h5ad"]                # Preferred file formats, ordered by preference
+    encodings   : Optional[List[str]] = ["dense"]     # Data encodings, e.g., "csr", "png"
+    validator   : Optional[str] = None                # Python function for validation (python:module:function)
+    validation_schema : Optional[Dict[str, Any]] = None # Optional schema for validator
+    description : Optional[str] = None                # Brief description
+
+    @property
+    def key(self) -> str:
+        """Canonical identifier used everywhere else."""
+        return f"{self.formats[0]}{self.modality}:{self.semantic}:{self.version}"
+```
+
+
+### `DatasetSpec`
+
+
+```python
+class DatasetSpec(BaseModel, extra="forbid"):
+    """
+    Describes a dataset's structure, provenance, and associated specifications.
+    """
+    name         : str                                 # Human-readable dataset name
+    description  : Optional[str] = None                # Brief description of the dataset
+    specs        : List[InputOutputSpecs]           # Unique set of input/output specifications
+    provenance   : Optional[Provenance] = None         # Dataset origin and authorship info
+    checksum     : Optional[str] = None                # Optional data integrity checksum
+    created_at   : Optional[datetime] = None           # Creation timestamp
+    metadata     : Dict[str, Any] = {}                 # Arbitrary additional metadata
+```
+
+
+###  `ModelSpec`
+
+
+```python
+class ModelSpec(BaseModel, extra="forbid"):
+    """
+    Describes a runnable model, its inputs, outputs, and metadata.
+    """
+    name              : str                                 # Human-readable model name
+    description       : Optional[str] = None                # Brief description of the model
+    version           : str                                 # Model version
+    required_inputs   : List[InputOutputSpecs] = []      # All required input specifications
+    optional_inputs   : List[InputOutputSpecs] = []      # Optional input specifications (e.g., masks)
+    outputs           :  List[InputOutputSpecs]  
+    compute           : Resource = Resource()               # Compute resource requirements
+    provenance        : Optional[Provenance] = None         # Model origin and authorship info
+    metadata          : Dict[str, Any] = {}                 # Arbitrary additional metadata
+    supported_tasks   : Optional[List[str]] = None          # List of supported TaskSpec names/IDs
```

Comment:
Please see comment here regarding supported_tasks and supported_metrics: https://github.com/chanzuckerberg/cz-benchmarks/pull/281#issuecomment-3000802864